### PR TITLE
Charlesmchen/prod fail

### PR DESCRIPTION
**Pull request by charlesmchen-signal**
I didn't apply these changes, since I think we should discuss exactly how this works.

PTAL @michaelkirk

### CODE REVIEW

File: SignalServiceKit/src/Util/OWSAsserts.h [hunk 1]
```diff
@@ -139,6 +139,18 @@ NS_ASSUME_NONNULL_BEGIN
...
+    do {                                                                                                               \
+        OWSFail(_messageFormat, ##__VA_ARGS__);                                                                        \
+        exit(0);                                                                                                       \
+    } while (0)
+
+#define OWSCProdExit(_messageFormat, ...)                                                                              \
+    do {                                                                                                               \
+        OWSCFail(_messageFormat, ##__VA_ARGS__);                                                                       \
+        exit(0);                                                                                                       \
+    } while (NO)
```
Comment (charlesmchen-signal) [inline SignalServiceKit/src/Util/OWSAsserts.h]: Options:

* We could add these new `OWSExit()` macros and then migrate (most?) `OWSFail()` macros to use them.  
* Same, but we could call them `OWSProdFail()` and rename the existing analytics macros with those names to something else.
* We could simply call the exiting macros `OWSFail()` and rename the existing macros to `OWSFailWithoutExit()` or whatever.  WDYT?

File: SignalServiceKit/src/Util/OWSSwiftUtils.swift [hunk 2]
```diff
@@ -15,8 +15,41 @@ public func assertOnQueue(_ queue: DispatchQueue) {
...
+// Once we're on Swift4.2 we can mark this as inlineable
+// @inlinable
+public func owsProdFail(_ rawMessage: String,
+                        file: String = #file,
+                        function: String = #function,
+                        line: Int = #line) {
+    let message = "\(file) \(function) \(line): \(rawMessage)"
+    Logger.error(message)
+    Logger.flush()
+    // TODO: Should we use this or fatalError()?
```
Comment (charlesmchen-signal) [inline SignalServiceKit/src/Util/OWSSwiftUtils.swift]: I haven't researched this question yet.

File: SignalServiceKit/src/Util/OWSSwiftUtils.swift [hunk 3]
```diff
@@ -15,8 +15,41 @@ public func assertOnQueue(_ queue: DispatchQueue) {
...
+                    line: Int = #line) {
+    let message = "\(file) \(function) \(line): \(rawMessage)"
+    Logger.error(message)
+    Logger.flush()
+    assertionFailure(message)
+}
+
+// Once we're on Swift4.2 we can mark this as inlineable
+// @inlinable
+public func owsProdFail(_ rawMessage: String,
```
Comment (charlesmchen-signal) [inline SignalServiceKit/src/Util/OWSSwiftUtils.swift]: Presumably the naming of these methods will align with our decisions for the Obj-C macros.

File: SignalServiceKit/src/Util/OWSSwiftUtils.swift [hunk 4]
```diff
@@ -15,8 +15,41 @@ public func assertOnQueue(_ queue: DispatchQueue) {
...
     assertionFailure(message)
 }
+
+// Once we're on Swift4.2 we can mark this as inlineable
+// @inlinable
+public func owsFail(_ rawMessage: String,
+                    file: String = #file,
+                    function: String = #function,
+                    line: Int = #line) {
+    let message = "\(file) \(function) \(line): \(rawMessage)"
```
Comment (charlesmchen-signal) [inline SignalServiceKit/src/Util/OWSSwiftUtils.swift]: We'll actually format these messages using the `owsFormatLogMessage()` function defined in:

https://github.com/signalapp/Signal-iOS/pull/3842

**Comment by michaelkirk-signal** [comment]
It seems like these changes are also in https://github.com/signalapp/Signal-iOS/pull/3843?

I left some comments in that PR related to this. Can you separate them? or base one on the other so it's clear what I'm reviewing?

**Comment by charlesmchen-signal** [comment]
> It seems like these changes are also in #3843?

Sorry about that. PTAL @michaelkirk

File: SignalServiceKit/src/Util/OWSSwiftUtils.swift [hunk 5]
```diff
@@ -32,3 +26,27 @@ public func AssertIsOnMainThread(file: String = #file,
...
+    // TODO: Format using owsFormatLogMessage() once it is merged.
+    let message = "\(file) \(function) \(line): \(rawMessage)"
+    Logger.error(message)
+    Logger.flush()
+    assertionFailure(message)
+}
+
+// Once we're on Swift4.2 we can mark this as inlineable
+// @inlinable
+public func owsProdExit(_ rawMessage: String,
```
[the remark this answers is not in the document]
Comment (michaelkirk-signal) [reply SignalServiceKit/src/Util/OWSSwiftUtils.swift]: Swift actually provides a nice set of assertions for debug and production, but as far as I can tell there's no way to plug them into our logging.

File: SignalServiceKit/src/Util/OWSSwiftUtils.swift [hunk 6]
```diff
@@ -32,3 +26,27 @@ public func AssertIsOnMainThread(file: String = #file,
...
+                        file: String = #file,
+                        function: String = #function,
+                        line: Int = #line) {
+
+    owsFail(rawMessage, file: file, function: function, line: line)
+    fatalError()
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply SignalServiceKit/src/Util/OWSSwiftUtils.swift]: Done.

File: SignalServiceKit/src/Util/OWSSwiftUtils.swift [hunk 7]
```diff
@@ -32,3 +26,27 @@ public func AssertIsOnMainThread(file: String = #file,
...
 }
+
+// Once we're on Swift4.2 we can mark this as inlineable
+// @inlinable
+public func owsFail(_ rawMessage: String,
+                    file: String = #file,
+                    function: String = #function,
+                    line: Int = #line) {
+    // TODO: Format using owsFormatLogMessage() once it is merged.
+    let message = "\(file) \(function) \(line): \(rawMessage)"
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply SignalServiceKit/src/Util/OWSSwiftUtils.swift]: Yeah this is just temporary until `owsFormatLogMessage()` is merged in the logging PR.

File: SignalServiceKit/src/Util/OWSSwiftUtils.swift (hunk 4 above)
Comment (charlesmchen-signal) [reply SignalServiceKit/src/Util/OWSSwiftUtils.swift]: Done.

File: SignalServiceKit/src/Util/OWSSwiftUtils.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply SignalServiceKit/src/Util/OWSSwiftUtils.swift]: Great question. It turns out we can return `-> Never`.  I've resurrected `notImplemented()` to take advantage of this and applied this in the "exiting" Swift fail.

File: SignalServiceKit/src/Util/OWSAsserts.h (hunk 1 above)
Comment (charlesmchen-signal) [reply SignalServiceKit/src/Util/OWSAsserts.h]: Done, with `OWSFailDebug()`.

**Comment by michaelkirk-signal** [comment]
>  charlesmchen-signal closed this an hour ago

Was this merged? I don't see the commit.

**Comment by charlesmchen-signal** [comment]
@michaelkirk do you see it now?

77711df27

**Comment by michaelkirk** [comment]
Yes, thanks!

**Pull request closed without merging**